### DISCUSSION TITLE
HA integration needed please!  :)

### DISCUSSION THREAD

**Discussion by dariusz737** [dzurikmiroslav/esp32-evse#70] [General]
Finally I have finish two units, for two cars. But desperately need to make Home Assistant integration. I need minimum status, charging on, off, charging power settings. Description in wiki are black magic for me. I have integrated few devices so far, but I have no point to start here. Please someone help me?

**Comment by dzurikmiroslav** [comment]
I haven't had time to learn how to work with HA yet :)
Do you need to integrate with the MQTT broker?

**Comment by dariusz737** [comment]
Yes. What adres should I put, to send or receive? 
<img width="609" alt="Zrzut ekranu 2024-09-25 o 00 12 55" src="https://github.com/user-attachments/assets/5c3a92f2-72ea-4ede-b253-b51f81047a8b">

**Reply by dariusz737** [reply]
What is topic, how to sen data to EVSE, to receive, I have four solution and topic is "ladowarka-lewa" in my case, I have found this and changed in this line -   client:publish("ladowarka-lewa", json.encode(payload)).

**Reply by dzurikmiroslav** [reply]
First of all, I recommend reading a some of articles about HA and MQTT.
You need deploy MQTT broker or use public broker. EVSE and HA must be connected to same broker, they subscribe and publish messages through the topic. Yes, via `client:publish` it will be sent message with the given topic.

**Reply by dariusz737** [reply]
I was reading, all I found reflects to Shelby etc... :)  I am receiving packets (look at screenshot) , and I will learn how to proces them and change fot "entities". Bu what to put in "topic" and "payload" in the upper part of screenshot to change current to 10A.
<img width="571" alt="Zrzut ekranu 2024-09-25 o 10 32 39" src="https://github.com/user-attachments/assets/05605825-9fc5-4b3d-95f5-1cd1a314e4b2">

**Reply by dariusz737** [reply]
I have broker, I can receive message via topic "ladowarka-lewa" as I have named it. How to send it, via what topic, what to put in payload.

**Reply by dariusz737** [reply]
I trying to deal with this, no success. for sure wrong syntax. How to receive message, an set current for example via MQTT :) 
<img width="507" alt="Zrzut ekranu 2024-09-25 o 10 54 59" src="https://github.com/user-attachments/assets/e74682ef-f27f-42f0-bd19-6723273ccc70">

**Reply by dzurikmiroslav** [reply]
Yes, I forgot to make MQTT subscribe script example in the wiki :)

Here is sample which I recently test:
```
local mqtt = require("mqtt")

local client = mqtt.client("mqtt://broker.hivemq.com:1883")

client:setonmessage(function(topic, data)
  if topic == "ladowarka-lewa/setcurrent" then
    evse.setchargingcurrent(tonumber(data))
  end
end)

client:setonconnect(function()
  print("MQTT contected")
  client:subscribe("ladowarka-lewa/setcurrent")
end)

client:connect()
```
And publish this message:
![image](https://github.com/user-attachments/assets/089c2cd7-e6fd-47a6-a5f8-a7de317c4a12)

**Reply by dariusz737** [reply]
Thank you very much. Finally works a little :) Small problem with f... dictionary, wich changed "l" to "ł" and I didn't notice... 
But for every command I have to set "setonmessage"? How to make it working as on first example with json ... 

I don't know from where I have got this... it was two weeks ago :) 

client:setonmessage(function(topic, message)
  print("Received message from topic: " .. topic)
  local data = json.decode(message)
  -- Process received data here
  if data.method == "setCurrent" then
    evse.setchargingcurrent(data.params.current)
  end
  -- Add more processing logic as needed
end)

**Comment by dzurikmiroslav** [comment]
You mean like this?

```
local mqtt = require("mqtt")
local json = require("json")

local client = mqtt.client("mqtt://broker.hivemq.com:1883")

client:setonmessage(function(topic, data)
  if topic ==  "ladowarka-lewa/cmd" then
    local msg = json.decode(data)
    if msg.method == "setCurrent" then
      evse.setchargingcurrent(msg.param)
    end
  end
end)

client:setonconnect(function()
  print("MQTT contected")
  client:subscribe("ladowarka-lewa/cmd")
end)

client:connect()
```
![image](https://github.com/user-attachments/assets/43d140ae-77e9-4b78-9541-8e6cc36f4dad)

**Reply by dariusz737** [reply]
Thank you for prompt answer. Finally I have same progress. Why 10.1, not 10 :) , and how to send boolean to enable and disable charging. And how to send data from EVSE with "push" not periodically? Its so easy for you, I hope not to bothering you too much.

**Reply by dariusz737** [reply]
Ok. I have found answer for most questions. Expect one, how to "push" data on change. :)

**Reply by dzurikmiroslav** [reply]
Just manually test the values to static variables if a change and then publish new values. I don't know a better way.

**Reply by dariusz737** [reply]
Only to find out correct code :D

**Comment by dariusz737** [comment]
So, I am on the way with integration... I have already sensors, state, etc.. Will publish all after I will sort it out. Please let me know what does mean all states A1, B2, etc. I can't find explanation anywhere in docs.

**Reply by dzurikmiroslav** [reply]
They are states according to the J1772 specification https://en.wikipedia.org/wiki/SAE_J1772
The expansion of states according to the suffix 1-2 is inspired by the phoenixcontact AC charging controller:
![image](https://github.com/user-attachments/assets/1d913783-b259-4cc3-894d-cd40cf25598f)
(source: https://www.phoenixcontact.com/en-pc/products/ac-charging-controller-ev-cc-ac1-m3-cc-ser-pcb-1622460)

I think that for the needs of HA integration it is essential to know whether the EV is connected and whether it is charging:
```
  local payload = {}
  local state = evse.getstate()
  payload["session"] = state >= evse.STATEB1 and state <= evse.STATED2
  payload["charging"] = state == evse.STATEC2 or state == evse.STATED2
```

**Reply by dzurikmiroslav** [reply]
So that I don't forget `evse.STATE{XX}` constants copy the enumerator https://github.com/dzurikmiroslav/esp32-evse/blob/v0.10.1/components/evse/include/evse.h#L26, in Lua it is a numerical constant, so numerical comparation can be used...

**Reply by dariusz737** [reply]
Unfortunately this returns always 0 
  payload["charging"] = (state == evse.STATEC2 or state == evse.STATED2) and 1 or 0
I have changed it for:
  local charging = (state == 4) and 1 or 0
I don't see relation between STATE, and number returned by - evse.getstate() 
Any hint please :) ?

**Reply by dzurikmiroslav** [reply]
Can you provide the whole code?

**Comment by dariusz737** [comment]
Finally progress... However still a lot to do. I need to change NEXION for my needs. And I belive there is some issues with stability. Maybe worth to consider some kind of watchdog, and forced minimum charging current after restart? When will be ready I can add my HA integration to hour project with your help. I have GitHub, but not much familiar with it. 
<img width="381" alt="Zrzut ekranu 2024-09-29 o 14 53 23" src="https://github.com/user-attachments/assets/08cc7775-98fa-4664-af66-7e80cd048085">

**Reply by Mecmohammed** [reply]
> أخيرًا، تحقق تقدم... لكن لا يزال هناك الكثير لأفعله. أحتاج إلى تعديل NEXION بما يتناسب مع احتياجاتي. وأعتقد أن هناك بعض المشاكل في الاستقرار. ربما يجدر التفكير في استخدام نوع من المراقبة، وفرض حد أدنى لتيار الشحن بعد إعادة التشغيل؟ متى سيكون جاهزًا؟ يمكنني إضافة تكامل HA الخاص بي إلى مشروع الساعة بمساعدتكم. لديّ حساب على GitHub، لكنني لستُ على دراية به. <img alt="زرزوت الكتروني 2024-09-29 o 14 53 23" width="381" src="https://private-user-images.githubusercontent.com/81053092/371855773-08cc7775-98fa-4664-af66-7e80cd048085.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDI3MDM2MTYsIm5iZiI6MTc0MjcwMzMxNiwicGF0aCI6Ii84MTA1MzA5Mi8zNzE4NTU3NzMtMDhjYzc3NzUtOThmYS00NjY0LWFmNjYtN2U4MGNkMDQ4MDg1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzIzVDA0MTUxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE1Mjg4NWI2NDM2NTcwMWVmMDI1NjM4NTdhOGVjMjQ4YzMxYjFlNTUzZWNjYjI1NjhkODc4MTY2ZTllODBiNjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.k2FIjme-dUxNdi3sE8ONIeraaJ5ocs2MWZDkYV-PDz4">

 hi  Can you tell me how you achieved this?

**Comment by Radwan-Alalili** [comment]
Hello there, I hope you doing well, you did a great work, thanks.

But I just want to know if I choose the small version which is the esp32 single phase, does the code work or it need some changes.

**Reply by dzurikmiroslav** [reply]
Yes, as I write in the README _One firmware to rule them all_ :-) Everything else is configurable via `board.cfg` config.
You can use the web installer and then edit the `board.cfg` via the Web UI.

**Reply by Radwan-Alalili** [reply]
Ok, thank you for your support. I will try to do this project and if I have any problems I will reach you here, but I see in the box of the Ev charger which you build there are other device that connected to the motherboard like emergency bottom, the electric connectors and other things, is there an electric diagram that shows how to connect them.

**Reply by dzurikmiroslav** [reply]
My wallbox wiring is described here https://dzurikmiroslav.github.io/esp32-evse-docs/10-hardware/esp32s2-evse-d-a/#with-socket-outlet
The emergency button is connected to the `IN2` switch 12V to input, here is the driver which handles it:
```
local aux = require("aux")

return {
  name = "Emergency stop",
  every1s = function ()
    local emstop = aux.read("IN2")
    if not emstop ~= evse.getavailable() then
      evse.setavailable(not emstop)
    end
  end
}
```

**Comment by taHC81** [comment]
Integrated within HA using MQTT, here's my card. Using the HA automations it changes the EV charging current to keep selected battery minimal charge rate. It can also temporary disable EV charge and then enable when conditions are met.

![image](https://github.com/user-attachments/assets/00d6485b-a0fc-454f-bd5f-9eadb1e0b0a8)

**Comment by nagyrobi** [comment]
> But desperately need to make Home Assistant integration. 

Also possible with: https://github.com/nagyrobi/esp32-evse_esphome-lvgl

**Comment by nagyrobi** [comment]
Lua script example: https://dzurikmiroslav.github.io/esp32-evse-docs/20-software/Lua/#home-assistant-mqtt-with-autodiscovery
